### PR TITLE
api: Add user permission endpoints

### LINKED ISSUE #310: api: Create endpoints to grand, revoke and list user permissions
Create the following endpoints to handle user permissions: 

- [ ] user.intent.grantPermission
- [ ] user.intent.revokePermission
- [ ] user.intent.listPermission

### PR DESCRIPTION
**Pull request by mathiashoeld**
# Changes
The following endpoints have been added:
- user.intent.grantPermission
- user.intent.revokePermission
- user.intent.listPermissions

Unit tests for 'user.intent.grantPermission' and 'user.intent.grantPermission' have been added.

Closes #310

### CODE REVIEW

File: api/src/authz/intents.ts [hunk 1]
```diff
@@ -68,6 +71,9 @@ export const globalIntents: Intent[] = [
   "global.createGroup",
   "user.authenticate",
   "user.changePassword",
+  "user.intent.listPermissions",
+  "user.intent.grantPermission",
+  "user.intent.revokePermission",
```
Comment (kevinbader) [inline api/src/authz/intents.ts]: as discussed please remove this and the `user.changePassword` intent from the list of global intents

File: api/src/authz/intents.ts [hunk 2]
```diff
@@ -105,7 +111,14 @@ export const userDefaultIntents: Intent[] = [
...
 
-export const userIntents: Intent[] = ["user.view", "user.authenticate", "user.changePassword"];
+export const userIntents: Intent[] = [
+  "user.view",
+  "user.authenticate",
+  "user.changePassword",
+  "user.intent.listPermissions",
+  "user.intent.grantPermission",
+  "user.intent.revokePermission",
+];
```
Comment (kevinbader) [inline api/src/authz/intents.ts]: newline

File: api/src/service/domain/organization/user_permission_grant.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import { ServiceUser } from "./service_user";
+import * as Project from "../workflow/project";
+import * as UserRecord from "./user_record";
+import * as ProjectEventSourcing from "../workflow/project_eventsourcing";
+import * as UserEventSourcing from "./user_eventsourcing";
+import * as UserPermissionGranted from "./user_permission_granted";
+import { PreconditionError } from "../errors/precondition_error";
+
+interface Repository {
+  getUser(userId: UserRecord.Id): Promise<Result.Type<UserRecord.UserRecord>>;
```
Comment (kevinbader) [inline api/src/service/domain/organization/user_permission_grant.ts]: `getPermissions` instead?

File: api/src/service/domain/organization/user_permission_grant.spec.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+          permissions: { "user.intent.grantPermission": [alice.id] },
+        }),
+    });
+    if (Result.isErr(result)) {
+      throw result;
+    }
+
+    const sourcedUser = result.reduce(
+      (user, event) => newUserFromEvent(ctx, user, event),
+      baseUser,
```
Comment (kevinbader) [inline api/src/service/domain/organization/user_permission_grant.spec.ts]: not baseUser but user that is used in `getUser`

File: api/src/service/domain/organization/user_permission_grant.spec.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        Promise.resolve({
+          ...baseUser,
+          permissions: { "user.intent.grantPermission": [alice.id] },
+        }),
+    });
+    if (Result.isErr(result)) {
+      throw result;
+    }
+
+    const sourcedUser = result.reduce(
```
Comment (kevinbader) [inline api/src/service/domain/organization/user_permission_grant.spec.ts]: `sourcedUser` = `userAfterGrantingPermission`

File: api/src/service/domain/organization/user_permission_grant.spec.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+      (user, event) => newUserFromEvent(ctx, user, event),
+      baseUser,
+    );
+    if (Result.isErr(sourcedUser)) {
+      throw sourcedUser;
+    }
+
+    assert.isTrue(Result.isOk(result), "Alice is authorized to grant this permission");
+    assert.isTrue(result.length > 0, "An event is created");
```
Comment (kevinbader) [inline api/src/service/domain/organization/user_permission_grant.spec.ts]: instead: bob had no permission earlier; bob has permission now. perhaps: alice still has the permission.

File: api/src/service/domain/organization/user_permission_grant.spec.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    const sourcedUser = result.reduce(
+      (user, event) => newUserFromEvent(ctx, user, event),
+      testUser,
+    );
+    if (Result.isErr(sourcedUser)) {
+      throw sourcedUser;
+    }
+
+    assert.isTrue(Result.isOk(result), "Alice is authorized to grant this permission");
+    assert.deepEqual(result, []);
```
Comment (kevinbader) [inline api/src/service/domain/organization/user_permission_grant.spec.ts]: in addition: Alice still has the permission

File: api/src/service/domain/organization/user_permission_revoke.spec.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  encryptedPrivKey: "encrypted",
+  log: [],
+  additionalData: {},
+};
+
+const baseRepository = {
+  getUser: async () => baseUser,
+};
+
+describe("Revoking user permissions: permissions", () => {
```
Comment (kevinbader) [inline api/src/service/domain/organization/user_permission_revoke.spec.ts]: same here

File: api/src/authz/intents.ts (hunk 1 above)
Comment (mathiashoeld) [reply api/src/authz/intents.ts]: remoed

File: api/src/authz/intents.ts (hunk 2 above)
Comment (mathiashoeld) [reply api/src/authz/intents.ts]: inserted

File: api/src/service/domain/organization/user_permission_grant.ts (hunk 3 above)
Comment (mathiashoeld) [reply api/src/service/domain/organization/user_permission_grant.ts]: good idea! but how should we then validate the event? in [line 58](https://github.com/openkfw/TruBudget/pull/311/files/edca4ce6f95bdea6eb703af7a361ec0315dabd01#diff-59fa7b2a05c74650348ac63bc68e30feR58) we perform event sourcing on this user to check the validity.

File: api/src/service/domain/organization/user_permission_grant.spec.ts (hunk 4 above)
Comment (mathiashoeld) [reply api/src/service/domain/organization/user_permission_grant.spec.ts]: put user in separate variable `permissionTestUser` and used that one for sourcing

File: api/src/service/domain/organization/user_permission_grant.spec.ts (hunk 5 above)
Comment (mathiashoeld) [reply api/src/service/domain/organization/user_permission_grant.spec.ts]: done

File: api/src/service/domain/organization/user_permission_grant.spec.ts (hunk 6 above)
Comment (mathiashoeld) [reply api/src/service/domain/organization/user_permission_grant.spec.ts]: done with the following assertions: 
```js
    // Bob now has the permission
    assert.isTrue(
      userAfterGrantingPermission.permissions["user.intent.grantPermission"]!.some(
        x => x === bob.id,
      ),
    );
    // Alice still has the permission
    assert.isTrue(
      userAfterGrantingPermission.permissions["user.intent.grantPermission"]!.some(
        x => x === alice.id,
      ),
    );
```

File: api/src/service/domain/organization/user_permission_revoke.spec.ts (hunk 8 above)
Comment (mathiashoeld) [reply api/src/service/domain/organization/user_permission_revoke.spec.ts]: done

File: api/src/service/domain/organization/user_permission_grant.spec.ts (hunk 7 above)
Comment (mathiashoeld) [reply api/src/service/domain/organization/user_permission_grant.spec.ts]: implemented by using 
```js
// Alice still has the permission
    assert.isTrue(
      userAfterGrantingPermission.permissions["user.intent.grantPermission"]!.some(
        x => x === alice.id,
      ),
    );
```

**Pull request merged**